code/modules/power/batteryrack.dm

```diff
@@ -1,234 +1,313 @@
+	var/ui_tick = 0
+
 
 /obj/machinery/power/smes/batteryrack/New()
 	..()
+	add_parts()
 	RefreshParts()
-	return
 
-//Maybe this should be moved up to obj/machinery
 /obj/machinery/power/smes/batteryrack/proc/add_parts()
 	component_parts = list()
-	component_parts += new /obj/item/weapon/cell/high
-	component_parts += new /obj/item/weapon/cell/high
-	component_parts += new /obj/item/weapon/cell/high
-	return
+	component_parts += new /obj/item/weapon/circuitboard/batteryrack
+	component_parts += new /obj/item/weapon/stock_parts/capacitor/				// Capacitors: Maximal I/O
+	component_parts += new /obj/item/weapon/stock_parts/capacitor/
+	component_parts += new /obj/item/weapon/stock_parts/capacitor/
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin/				// Matter Bin: Max. amount of cells.
 
 
 /obj/machinery/power/smes/batteryrack/RefreshParts()
-	capacitors_amount = 0
-	cells_amount = 0
-	var/max_level = 0 //for both input and output
+	var/capacitor_efficiency = 0
+	var/maxcells = 0
 	for(var/obj/item/weapon/stock_parts/capacitor/CP in component_parts)
-		max_level += CP.rating
-		capacitors_amount++
-	input_level_max = 50000 + max_level * 20000
-	output_level_max = 50000 + max_level * 20000
+		capacitor_efficiency += CP.rating
 
-	var/C = 0
-	for(var/obj/item/weapon/cell/PC in component_parts)
-		C += PC.maxcharge
-		cells_amount++
-	capacity = C * 40   //Basic cells are such crap. Hyper cells needed to get on normal SMES levels.
+	for(var/obj/item/weapon/stock_parts/matter_bin/MB in component_parts)
+		maxcells += MB.rating * 3
 
+	max_transfer_rate = 10000 * capacitor_efficiency // 30kw - 90kw depending on used capacitors.
+	max_cells = min(PSU_MAXCELLS, maxcells)
+	input_level = max_transfer_rate
+	output_level = max_transfer_rate
+
+/obj/machinery/power/smes/batteryrack/Destroy()
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		qdel(C)
+	internal_cells = null
+	return ..()
 
 /obj/machinery/power/smes/batteryrack/update_icon()
 	overlays.Cut()
-	if(stat & BROKEN)	return
-
-	if(!br_cache)
-		br_cache = list()
-		br_cache.len = 7
-		br_cache[1] = image('icons/obj/power.dmi', "gsmes_outputting")
-		br_cache[2] = image('icons/obj/power.dmi', "gsmes_charging")
-		br_cache[3] = image('icons/obj/power.dmi', "gsmes_overcharge")
-		br_cache[4] = image('icons/obj/power.dmi', "gsmes_og1")
-		br_cache[5] = image('icons/obj/power.dmi', "gsmes_og2")
-		br_cache[6] = image('icons/obj/power.dmi', "gsmes_og3")
-		br_cache[7] = image('icons/obj/power.dmi', "gsmes_og4")
-
-	if (output_attempt)
-		overlays += br_cache[1]
-	if(inputting)
-		overlays += br_cache[2]
-
-	var/clevel = chargedisplay()
-	if(clevel>0)
-		overlays += br_cache[3+clevel]
-	return
+	icon_update = 0
+
+	var/cellcount = 0
+	var/charge_level = between(0, round(Percentage() / 12), 7)
+
+
+	overlays += "charge[charge_level]"
+
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		cellcount++
+		overlays += "cell[cellcount]"
```

Comment:
Should this be checking against PSU_MAXCELLS?

Reply:
Unnecessary, as PSU_MAXCELLS is also being checked when trying to insert a new cell. Therefore it is not possible to exceed 9 cells even with T4/T5 components.